### PR TITLE
Add automatic Documentation

### PR DESCRIPTION
**Pull request by harrypuuter**
This PR adds automatic documentation via doxygen + sphinx to the repository. The documentation can be updated by using the `-DRUN_DOCS` option with cmake. 

The procedure is adapted from https://devblogs.microsoft.com/cppblog/clear-functional-c-documentation-with-sphinx-breathe-doxygen-cmake/

### CODE REVIEW

**Comment by harrypuuter** [comment]
The automatic triggering of sphinx does not work as intended yet i think

**Comment by harrypuuter** [comment]
now the paths should be correct for the automatic github pages as soon as this is merged to master

File: docgenerator/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,40 @@
+find_package(Doxygen REQUIRED)
+find_package(Sphinx REQUIRED)
+
+# adapted from https://devblogs.microsoft.com/cppblog/clear-functional-c-documentation-with-sphinx-breathe-doxygen-cmake/
+# Find all headers
+file(GLOB_RECURSE HEADERS ${PROJECT_SOURCE_DIR}/*.hxx)
+
+set(DOXYGEN_INPUT_DIR ${PROJECT_SOURCE_DIR}/)
+set(DOXYGEN_OUTPUT_DIR ${PROJECT_SOURCE_DIR}/docs/doxygen/)
```
Comment (stwunsch) [inline docgenerator/CMakeLists.txt:9]: The output dir must be in the build directory. Never touch the source directory!

File: docgenerator/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+set(DOXYGEN_INDEX_FILE ${DOXYGEN_OUTPUT_DIR}/html/index.html)
+set(DOXYFILE_IN ${PROJECT_SOURCE_DIR}/docgenerator/scripts/Doxyfile.in)
+set(DOXYFILE_OUT ${PROJECT_SOURCE_DIR}/docgenerator/scripts/Doxyfile)
+
+#Replace variables inside @@ with the current values
+configure_file(${DOXYFILE_IN} ${DOXYFILE_OUT} @ONLY)
+
+file(MAKE_DIRECTORY ${DOXYGEN_OUTPUT_DIR}) #Doxygen won't create this for us
+add_custom_command(OUTPUT ${DOXYGEN_INDEX_FILE}
```
Comment (stwunsch) [inline docgenerator/CMakeLists.txt:18]: All of these targets are at build time, so when calling `make`, not `cmake` (regarding the comment above).

File: docgenerator/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                   DEPENDS ${HEADERS}
+                   COMMAND ${DOXYGEN_EXECUTABLE} ${DOXYFILE_OUT}
+                   MAIN_DEPENDENCY ${DOXYFILE_OUT} ${DOXYFILE_IN}
+                   COMMENT "Generating docs")
+
+add_custom_target(Doxygen ALL DEPENDS ${DOXYGEN_INDEX_FILE})
+
+set(SPHINX_SOURCE ${PROJECT_SOURCE_DIR}/docgenerator/scripts/sphinx_source/)
```
Comment (stwunsch) [inline docgenerator/CMakeLists.txt:26]: All to build dir.

File: docgenerator/scripts/Doxyfile [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#...
+INPUT = "/work/sbrommer/ntuple_prototype/"
```
Comment (stwunsch) [inline docgenerator/scripts/Doxyfile:2]: This file should not be commited.

File: docgenerator/scripts/sphinx_source/Makefile [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# Minimal makefile for Sphinx documentation
```
Comment (stwunsch) [inline docgenerator/scripts/sphinx_source/Makefile:1]: Why do we need a makefile and not just put this as a custom command in cmake?

File: docgenerator/scripts/sphinx_source/conf.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+
+# -- Project information -----------------------------------------------------
+
+project = "ToBeNamedPrototype"
+copyright = "2021, Sebastian Brommer, Maximilian Burkart, Artur Gottmann, Sebastian Wozniewski, Stefan Wunsch"
+author = "Sebastian Brommer, Maximilian Burkart, Artur Gottmann, Sebastian Wozniewski, Stefan Wunsch"
+
+# The full version, including alpha/beta/rc tags
+release = "0.1"
```
Comment (stwunsch) [inline docgenerator/scripts/sphinx_source/conf.py:25]: This should be a variable in cmake, at some point and then configured via cmake.

File: docgenerator/scripts/sphinx_source/conf.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+# relative to this directory. They are copied after the builtin static files,
+# so a file named "default.css" will overwrite the builtin "default.css".
+html_static_path = ["_static"]
+
+# ...
+extensions = ["breathe"]
+# ...
+
+# Breathe Configuration
+breathe_default_project = "ToBeNamedPrototype"
```
Comment (stwunsch) [inline docgenerator/scripts/sphinx_source/conf.py:61]: You can also template the project name!

File: docgenerator/scripts/sphinx_source/index.rst [hunk 8]
```diff
@@ -0,0 +1,32 @@
+.. a documentation master file, created by
+   sphinx-quickstart on Thu Apr 29 16:29:35 2021.
+   You can adapt this file completely to your liking, but it should at least
+   contain the root `toctree` directive.
+
+Welcome to The FARTUS documentation!
```
Comment (stwunsch) [inline docgenerator/scripts/sphinx_source/index.rst:6]: FARTUS? :D

File: docgenerator/scripts/sphinx_source/make.bat [hunk 9]
```diff
@@ -0,0 +1,35 @@
+@ECHO OFF
+
+pushd %~dp0
+
+REM Command file for Sphinx documentation
```
Comment (stwunsch) [inline docgenerator/scripts/sphinx_source/make.bat:5]: I think we can drop windows support ;)

File: docs/.buildinfo [hunk 10]
```diff
@@ -0,0 +1,4 @@
+# Sphinx build info version 1
```
Comment (stwunsch) [inline docs/.buildinfo:1]: Remove this file

Comment (stwunsch) [reply docs/.buildinfo:1]: And also all the other .`doctrees` files below.

**Comment by stwunsch** [comment]
Basically two things: There are many files which should not be commited and you should set the paths always to the build diretory for generated files or build artefacts.

**Comment by stwunsch** [comment]
I've refactored your impl in #16. I think it's best to make this a separate cmake project, because there's no compilation or similar part of the make step and simplifies the CI later.

**Comment by harrypuuter** [comment]
I think your refactored version is so much better than this one, i close this PR

**Pull request closed without merging**